packages/fxa-auth-server/lib/payments/configuration/manager.ts

```diff
@@ -0,0 +1,180 @@
+  private planConfigDbRef: TypedCollectionReference<PlanConfig>;
+  private log: Logger;
+  private prefix: string;
+  private cancelProductListener: (() => void) | undefined;
+  private cancelPlanListener: (() => void) | undefined;
+
+  private products: Record<string, ProductConfig> = {};
+  private plans: Record<string, PlanConfig> = {};
+
+  constructor() {
+    const config = Container.get(AppConfig);
+    this.prefix = `${config.authFirestore.prefix}payment-config-`;
+    this.firestore = Container.get(AuthFirestore);
+    this.log = Container.get(AuthLogger) as any as Logger;
+    this.productConfigDbRef = this.firestore.collection(
+      `${this.prefix}products`
+    ) as TypedCollectionReference<ProductConfig>;
+    this.planConfigDbRef = this.firestore.collection(
+      `${this.prefix}plans`
+    ) as TypedCollectionReference<PlanConfig>;
+  }
+
+  /**
+   * Load all products and plans from firestore.
+   *
+   * Note that this will overwrite any existing products and plans.
+   */
+  public async load() {
+    const [productResults, planResults] = await Promise.all([
+      this.productConfigDbRef.select().get(),
+      this.planConfigDbRef.select().get(),
+    ]);
+    productResults.docs.forEach((doc) => {
+      this.products[doc.id] = ProductConfig.fromFirestoreObject(
+        doc.data(),
+        doc.id
+      );
+    });
+    planResults.docs.forEach((doc) => {
+      this.plans[doc.id] = PlanConfig.fromFirestoreObject(doc.data(), doc.id);
+    });
+  }
+
+  /**
+   * Start the listeners for changes to the product and plan configs.
+   *
+   * Note that this will exit the process if it fails to start the listeners.
+   */
+  public startListeners() {
+    this.cancelProductListener = this.productConfigDbRef.onSnapshot(
+      (snapshot) => {
+        snapshot.docChanges().forEach((change) => {
+          if (['added', 'modified'].includes(change.type)) {
+            this.products[change.doc.id] = ProductConfig.fromFirestoreObject(
+              change.doc.data(),
+              change.doc.id
+            );
+          } else {
+            if (change.doc.id in this.products) {
+              delete this.products[change.doc.id];
+            }
+          }
+        });
+      },
+      (err) => {
+        this.log.error('startListener', { err });
+        process.exit(1);
+      }
+    );
+    this.cancelPlanListener = this.planConfigDbRef.onSnapshot((snapshot) => {
+      snapshot.docChanges().forEach(
+        (change) => {
+          if (['added', 'modified'].includes(change.type)) {
+            this.plans[change.doc.id] = PlanConfig.fromFirestoreObject(
+              change.doc.data(),
+              change.doc.id
+            );
+          } else {
+            if (change.doc.id in this.plans) {
+              delete this.plans[change.doc.id];
+            }
+          }
+        },
+        (err) => {
+          this.log.error('startListener', { err });
+          process.exit(1);
```

Comment:
Terminating the auth-server process seems really severe for this.  I worry that if there's a persistent error that we cannot solve quickly, this would effectively cause an outage.

Reply:
That's correct, there would be an outage anyways in this case since payments wouldn't work and we might otherwise miss a sentry/log error vs. not having the auth-server start at all. We've accidentally had broken payments before where we only noticed when a Mozillian or outside person reported they couldn't buy something. How soon we want to detect that and how alarming the response is something useful to consider. Log lines will be ignored, and our sentry has enough in it that we miss things in there now. It feels hard to miss auth-server refusing to start, that'll show up in the canary instance right away.

If this errors, we'll need to somehow code a solution that lets auth-server run in a degraded state where it might not be able to offer a checkout experience, similar to being unable to contact Stripe. What were your thoughts on this?

Reply:
My concern is having an FxA outage caused by a service (Firestore) and a library (google-cloud) that we don't have any control over.  I understand the need to know about the error.  IIUC the listener is for updating the docs, so subplat should continue to work, but the we cannot add new products, update plans, etc.?

I think at a minimum we need to verify how well FxA functions with `SUBSCRIPTIONS_ENABLED` toggled off.